python/pyxet/src/transactions.rs

```diff
@@ -0,0 +1,231 @@
+
+    /// This is an example for using doc comment attributes
+    static ref TRANSACTION_LIMIT_LOCK: Arc<Semaphore> = Arc::new(Semaphore::new((*MAX_NUM_CONCURRENT_TRANSACTIONS).load(Ordering::Relaxed)));
+}
+
+pub struct WriteTransaction {
+    transaction: Option<XetRepoWriteTransaction>,
+    branch: String,
+
+    pub new_files: Vec<String>,
+    pub copies: Vec<(String, String)>,
+    pub deletes: Vec<String>,
+    pub moves: Vec<(String, String)>,
+
+    // A transaction will be cancelled on completion unless
+    // this flag is set.
+    commit_when_ready: bool,
+
+    // For testing: generate an error on the commit to make sure everything works
+    // above this.
+    error_on_commit: bool,
+
+    // For tensting, things succeed but don't actually push.
+    do_not_commit: bool,
+
+    // For testing: go through everything, but don't actually do the last bit of the commit.
+    commit_canceled: bool,
+
+    // The message written on success
+    commit_message: String,
+    _transaction_permit: tokio::sync::OwnedSemaphorePermit,
+}
+
+impl WriteTransaction {
+    pub async fn new(
+        repo: &Arc<XetRepo>,
+        branch: &str,
+        commit_message: &str,
+    ) -> Result<Arc<RwLock<Self>>> {
+        let transaction = repo.begin_write_transaction(branch, None, None).await?;
+
+        debug!("PyWriteTransaction::new_transaction(): Acquiring transaction permit.");
+        let transaction_permit = TRANSACTION_LIMIT_LOCK.clone().acquire_owned().await?;
+
+        let write_transaction = Self {
+            transaction: Some(transaction),
+            branch: branch.to_string(),
+            commit_message: commit_message.to_string(),
+            copies: Vec::new(),
+            deletes: Vec::new(),
+            moves: Vec::new(),
+            new_files: Vec::new(),
+            do_not_commit: false,
+            error_on_commit: false,
+            commit_when_ready: false,
+            _transaction_permit: transaction_permit,
+            commit_canceled: false,
+        };
+
+        Ok(Arc::new(RwLock::new(write_transaction)))
+    }
+
+    /// Complete the transaction by either cancelling it or committing it, depending on flags.
+    pub async fn complete(&mut self) -> Result<()> {
+        if let Some(transaction) = self.transaction.take() {
+            if self.error_on_commit {
+                return Err(anyhow!("Error on commit flagged; Cancelling."));
+            }
+
+            if self.commit_canceled || !self.commit_when_ready {
+                info!("PyWriteTransactionInternal::complete: Cancelling commit.");
+                transaction.cancel().await?;
+            } else {
+                if !self.do_not_commit {
+                    transaction.commit(&self.commit_message).await?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn set_commit_when_ready(&mut self, commit_when_ready: bool) -> Result<()> {
```

Comment:
This function (and several similar others) changed to the below format since the default flags are always `false` and we only call this function to set to `true`.

```
pub fn set_commit_when_ready(&mut self) {
        self.commit_when_ready = true;
    }
```